spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/repository/query/GqlDatastoreQuery.java

```diff
@@ -0,0 +1,327 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.datastore.repository.query;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.google.cloud.Timestamp;
+import com.google.cloud.datastore.Blob;
+import com.google.cloud.datastore.Cursor;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.GqlQuery;
+import com.google.cloud.datastore.GqlQuery.Builder;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.Query.ResultType;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+
+import org.springframework.cloud.gcp.data.datastore.core.DatastoreOperations;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreMappingContext;
+import org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntity;
+import org.springframework.data.repository.query.Parameter;
+import org.springframework.data.repository.query.Parameters;
+import org.springframework.data.repository.query.QueryMethod;
+import org.springframework.data.repository.query.QueryMethodEvaluationContextProvider;
+import org.springframework.data.repository.query.RepositoryQuery;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ParserContext;
+import org.springframework.expression.common.CompositeStringExpression;
+import org.springframework.expression.common.LiteralExpression;
+import org.springframework.expression.spel.standard.SpelExpression;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.util.StringUtils;
+
+/**
+ * Query Method for GQL queries.
+ *
+ * @author Chengyuan Zhao
+ *
+ * @since 1.1
+ */
+public class GqlDatastoreQuery<T> implements RepositoryQuery {
+
+	private static final Map<Class<?>, Function<Builder, BiFunction<String, Object, Builder>>>
+			GQL_PARAM_BINDING_FUNC_MAP;
+
+	// A small string that isn't used in GQL syntax
+	private static String ENTITY_CLASS_NAME_BOOKEND = "|";
+
+	static {
+		GQL_PARAM_BINDING_FUNC_MAP = ImmutableMap
+				.<Class<?>, Function<Builder, BiFunction<String, Object, Builder>>>builder()
+				.put(Cursor.class, builder -> (s, o) -> builder.setBinding(s, (Cursor) o))
```

Comment:
Should we also map the array types like `String[]`?

Reply:
I'm thinking let's only support these singlular types directly right now (w/o conversion). I will make an issue for using conversion and arrays for a separate PR.

Reply:
You wouldn't need to convert `String[]`.

Reply:
@meltsufin changes made, PTAL